lib/p2p/Pool.ts

```diff
@@ -169,11 +171,28 @@ class Pool extends EventEmitter {
   private connectNodes = (nodes: NodeConnectionIterator, ignoreKnown = false, retryConnecting = false) => {
     const connectionPromises: Promise<void>[] = [];
     nodes.forEach((node) => {
-      // check that this node is not ourselves, that it has listening addresses,
+      const is_not_ignored = false;
+      let has_no_pending_connections = true;
+
+      // check that this node is not ourselves
+      const is_not_us = node.nodePubKey !== this.handshakeData.nodePubKey;
+
+      // that it has listening addresses,
+      const has_addresses = node.addresses.length > 0;
+
       // and that either we haven't heard of it, or we're not ignoring known nodes and it's not banned
-      if (node.nodePubKey !== this.handshakeData.nodePubKey && node.addresses.length > 0 &&
-        (!this.nodes.has(node.nodePubKey) || (!ignoreKnown && !this.nodes.isBanned(node.nodePubKey))) &&
-        !this.peers.has(node.nodePubKey)) {
+      const is_unknown_node = !this.nodes.has(node.nodePubKey);
+      if (!is_unknown_node) {
+        const is_not_ignored = !ignoreKnown && !this.nodes.isBanned(node.nodePubKey);
+
+        // Check we're not already trying to connect to this node.
+        if (node.nodePubKey in this.outgoingConnections) {
```

Comment:
I believe we want to do this check whether we "know" the node or not, right now I don't think this check would work if we have a pending connection to a node we've never connected to before.